### PR TITLE
don't attempt mempool entry for wallet transactions on startup if alr…

### PR DESCRIPTION
**Pull request by instagibbs**
…eady in mempool

Mempool loads first, wallet second. Second attempt fails, marking that transaction !fInMempool. Those funds will disappear until confirmation is reached.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Hmm, seems like its working around the issue instead of fixing, I think CWalletTx::AcceptToMemoryPool shouldn't unset fInMempool if its already set.

**Comment by MarcoFalke** [comment]
You could add the test to `mempool_persist.py` instead, as wallet is already a "test everything that has something to do with sending bitcoin"

**Comment by instagibbs** [comment]
@TheBlueMatt right, feels quite brittle, but wanted to throw up a working solution/test.

Knowing whether a boolean is already set sounds like boost::optional(?)

@MarcoFalke noted, will move

**Comment by instagibbs** [comment]
Well, I suppose you mean if it's set true, don't every change the value from that call.

**Comment by TheBlueMatt** [comment]
@instagibbs huh? No, just change the third-to-last line in wallet.cpp to fInMempool |= ret from fInMempool = ret.

**Comment by instagibbs** [comment]
@TheBlueMatt I honestly still don't understand the current comment in the logic. As long as this is only being called once on startup, your suggestion seems like a correct fix at least.

**Comment by Sjors** [comment]
> Knowing whether a boolean is already set sounds like boost::optional(?)

Or [std::optional](http://en.cppreference.com/w/cpp/utility/optional) as of C++17 :-)

**Comment by TheBlueMatt** [comment]
@instagibbs the entered-mempool callback referred to there is async, so you can create a transaction, put it in the mempool, and then the IsInMempool() call will return false for some time. Thus, change outputs will "not be spendable" until the callbacks catch up if we didnt set fInMempool to true in AcceptToMemoryPool.

**Comment by MarcoFalke** [comment]
```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Tested ACK f72f5e2576650c3fc314e16f33dbcf24e85520ec
-----BEGIN PGP SIGNATURE-----

iQIcBAEBCgAGBQJaKCkKAAoJENLqSFDnUoslHSkP/2QJqJa6aKFbp4ripP81rep5
jB074r5uPkh6Oa4MgbCI4dte9BQNkfIZ8wTl9sfNax50j8TMeglSHubn6P/F50A/
zo0HJG767dvQMxGd+KSxYAS9nk9Oh4iY6SZs44lVYvOxUjnzlRTl9ElbzmDW+BOY
deqwbH3xv25Ii9+xrvjzHBt/3jiWmdMDJ6NjqJofCmsrLARY3NHL/IQhAG8sqSPR
QSu81crg25IshaO55JPOu2EyXMhnv5M3AF3O4kqaxSFZ3CyioXJhMgggs+n/ZZbI
0hGOxNFHjAznTOFD3/a8/6ritVtpyhUbtPe00k0LcAssTWiqjLtfU2mEpF/HudYV
HvNdDogxXoXjsQTKjxtKYetm1oe/NrGtRpI207cMw5TL+q2XINRApgCJ48Sb5qVr
cOsVokhy1VdPS8OGQHGusiEfmUC60nX8thixbKbhf/z76NEFiPDNT4Nyq5qjIs82
ps2EbX9L66IPCKwP2w25L2EMK3cZaTkox2XNfQS0zKM/reDSzUSlypENA7V/R7dz
CaE7rCOwllHUu/mr9ANKcug3jeikDkrI7kFDg11Dxr9Pdt97YUrJtblYmL97Nlyp
NDZihNY1xCfFLUqQXwpxDma+VEcafeTw1R7mZDHWQwg04Vtd+U/EX+6Rw2Cs4TAU
awhmaenWj7pKls9ql2gW
=MeI2
-----END PGP SIGNATURE-----

**Comment by Sjors** [comment]
@TheBlueMatt although unrelated to this bug, isn't it a good idea in general to check `!mempool.exists(...)` before trying to add a transaction to the mempool? Or perhaps `AcceptToMemoryPool` should check if already contains the transaction before proceeding?

**Comment by MarcoFalke** [comment]
If you move the fix to `CWalletTx::AcceptToMemoryPool`, it should check for `mempool.exists()` instead of the `|=` logic. Imo this better documents by code what is being done.

**Comment by Sjors** [comment]
I clarified the steps to reproduce to say it may not be deterministic (though right now I'm having difficulty getting notifications to work at all, so recompiling v0.15.1 for comparison).

What's a good way to add debug statements to the various methods involved here? For one thing I'd like to rule out that OSX is somehow not showing the popup. See also [IRC](https://botbot.me/freenode/bitcoin-core-dev/2017-12-06/?msg=94312477&page=3).

**Comment by MarcoFalke** [comment]
@Sjors Probably posted that to the wrong thread

**Comment by TheBlueMatt** [comment]
I think |= is more correct than it is now anyway - if a transaction isn't in mempool and fails to be added, we shoul unset fInMempool in the callback, not in CWalletTx::AcceptToMemoryPool.

**Comment by instagibbs** [comment]
mark as 0.16 milestone?

**Comment by Sjors** [comment]
Merging this would make testing SegWit wallet changes in #11403 easier.

**Comment by instagibbs** [comment]
moved mempool check logic closer to source, moved test to mempool_persist.py

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -4114,6 +4114,11 @@ int CMerkleTx::GetBlocksToMaturity() const
 
 bool CWalletTx::AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState& state)
 {
+    // Quick check to avoid re-setting fInMempool to false
```
Comment (Sjors) [inline src/wallet/wallet.cpp]: And to avoid wasting time?

**Comment by laanwj** [comment]
utACK 6697a7089441deded836332e7ce3b4a1a9a3cbcd

**Comment by promag** [comment]
post-utACK.

**Pull request merged**